File: Problems/LeetCode/M-0300.py
from typing import List


class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        L = [1] * len(nums)
        for i in range(1, len(L)):
            subp = [L[k] for k in range(i) if nums[k] < nums[i]]
            L[i] = 1 + max(subp, default=0)
        return max(L, default=0)


nums = [10, 9, 2, 5, 3, 7, 101, 18]
    
f = Solution()
print(f.lengthOfLIS(nums))

File: Problems/LeetCode/M-0300.cpp
#include<bits/stdc++.h>

using namespace std;

class Solution {
public:
    int lengthOfLIS(vector<int> &nums) {
        vector<int> dp(nums.size(), 1); //init dp cache of length n with 1s
        for (int i = 0; i < nums.size(); ++i)
            for (int j = 0; j < i; ++j)
                if (nums[i] > nums[j] && dp[i] < dp[j] + 1)
                    dp[i] = dp[j] + 1;
        return *max_element(dp.begin(), dp.end());
    }
};

int main() {
    Solution s;
    vector<int> nums{10, 9, 2, 5, 3, 7, 101, 18};
    cout << s.lengthOfLIS(nums) << "\n";
    return 0;
}